File: 212-word-search-ii/word-search-ii.cpp
class Trie{
public:
    vector<Trie*> child= vector<Trie*>(26, NULL);
    bool isWord=false;

    Trie() {
        
    }
};
class Solution {
public:
    vector<string> ans;
    vector<int> dir= vector<int>({-1,0,1,0,-1});
    void insert(string word, Trie* curr) {
        for(int i=0;i<(int)word.size();++i){
            if(!curr->child[word[i]-'a']){
                curr->child[word[i]-'a']= new Trie();
            }
            curr= curr->child[word[i]-'a'];
        }
        curr->isWord=true;
    }
    void dfs(vector<vector<char>>& board, int r, int c, int m, int n, vector<vector<bool>> &vis, Trie* curr, string &str){
        if(curr && curr->isWord){
            ans.push_back(str);
            curr->isWord= false;
        } 
        if(!curr) return;
        if(r<0 || c<0 || r==m || c==n){
            return;
        }
        if(vis[r][c]) return;
        vis[r][c]=true;
        str.push_back(board[r][c]);
        if(curr->child[board[r][c]-'a']){
            for(int i=0;i<4;++i){
                int nr=r+dir[i], nc=c+dir[i+1];
                dfs(board, nr, nc, m, n, vis, curr->child[board[r][c]-'a'], str);
            }
        }
        str.pop_back();
        vis[r][c]=false;
        return;
    }
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        Trie* curr= new Trie();
        for(string word: words){
            insert(word, curr);
        }
        int m= (int)board.size(), n=(int)board[0].size();
        vector<vector<bool>> vis(m, vector<bool>(n));
        for(int i=0;i<m;++i){
            for(int j=0;j<n;++j){
                string str= "";
                dfs(board, i, j, m, n, vis, curr, str);
            }
        }
        return ans;
    }
};

File: 212-word-search-ii/word-search-ii.java
class TrieNode{
    TrieNode[] children;
    int isWord;
    public TrieNode(){
        children= new TrieNode[26];
        isWord= -1;
    }
}
class Solution {
    TrieNode root;
    int[] dir= {-1,0,1,0,-1};
    boolean[][] used= new boolean[12][12];
    Set<Integer> index= new HashSet<>();
    private void insert(String word, int idx) {
        TrieNode curr= root;
        for(char ch: word.toCharArray()){
            if(curr.children[ch-'a']==null){
                curr.children[ch-'a']= new TrieNode();
            }
            curr= curr.children[ch-'a'];
        }
        curr.isWord= idx;
    }
    private void checkIfPresent(int r, int c, char[][] board, TrieNode curr){
        if(curr.isWord!=-1){
            index.add(curr.isWord);
            curr.isWord=-1;
        }
        if(r<0 || r==board.length || c<0 || c==board[0].length || used[r][c]) return;
        char ch= board[r][c];
        if(curr.children[ch-'a']!=null){
            used[r][c]= true;
            for(int i=0;i<4;i++){
                checkIfPresent(r+dir[i],c+dir[i+1], board, curr.children[ch-'a']);
            }
            used[r][c]= false;
        }
        return;
    }
    public List<String> findWords(char[][] board, String[] words) {
        root= new TrieNode();
        for(int i=0;i<words.length;i++){
            insert(words[i], i);
        }
        for(int i=0;i<board.length;i++){
            for(int j=0;j<board[0].length;j++){
                checkIfPresent(i, j, board, root);
            }
        }
        List<String> ans= new ArrayList<>();
        for(int i: index){
            ans.add(words[i]);
        }
        return ans;
    }
}